### PR TITLE
refactor(protocol): optimize keyForName

### LINKED ISSUE #13290: feat: optimize keyForName method in AddressResolver.sol
### Describe the feature request

Changing AddressResolver's key generation from
```solidity
string memory key = string.concat(Strings.toString(chainId), ".", name);
```
to

```solidity
string memory key = string(bytes.concat(bytes32(chainId), bytes(name)));
```

will reduce 2360 gas for proveBlock tx. But it will break many hardhat tests.

| Deployment Cost                                | Deployment Size |        |        |        |         |
|------------------------------------------------|-----------------|--------|--------|--------|---------|
| proveBlock                                     | 49861           | 56789  | 49861  | 126161 | 49      |
| resolve(string,bool)(address)                  | 4329            | 4544   | 4554   | 10829  | 100     |
| resolve(uint256,string,bool)(address)          | 4232            | 4272   | 4232   | 6232   | 49      |

| Deployment Cost                                | Deployment Size |        |        |        |         |
|------------------------------------------------|-----------------|--------|--------|--------|---------|
| proveBlock                                     | 47501           | 54429  | 47501  | 123801 | 49      |
| resolve(string,bool)(address)                  | 3496            | 3709   | 3718   | 9996   | 100     |
| resolve(uint256,string,bool)(address)          | 3541            | 3581   | 3541   | 5541   | 49      |

### Describe alternatives you've considered

Description of the alternatives you've considered here.

### Additional context

You can work on top of this [draft PR](https://github.com/taikoxyz/taiko-mono/pull/13285) to do comparison by running:

```
forge test --gas-report -vvv
```

### PR DESCRIPTION
**Pull request by dantaik**
#### Previously
| Function Name                          | min             | avg    | median | max    | # calls |
|----------------------------------------|-----------------|--------|--------|--------|---------|
| keyForName                             | 2232            | 2323   | 2377   | 2377   | 40      |
| proposeBlock                           | 22385           | 32506  | 22405  | 92404  | 309     |
| proveBlock                             | 55304           | 65358  | 55304  | 131704 | 309     |
| resolve(string,bool)(address)          | 4407            | 4490   | 4407   | 10907  | 629     |
| resolve(uint256,string,bool)(address)  | 4310            | 4342   | 4310   | 6310   | 927     |
| verifyBlocks                           | 13575           | 14588  | 14293  | 80372  | 301     |

#### Now

| Function Name                          | min             | avg    | median | max    | # calls |
|----------------------------------------|-----------------|--------|--------|--------|---------|
| keyForName                             | 1459            | 1459   | 1459   | 1459   | 40      |
| proposeBlock                           | 21688           | 31809  | 21708  | 91707  | 309     |
| proveBlock                             | 52241           | 62295  | 52241  | 128641 | 309     |
| resolve(string,bool)(address)          | 3571            | 3654   | 3571   | 10071  | 629     |
| resolve(uint256,string,bool)(address)  | 3613            | 3646   | 3613   | 5616   | 927     |
| verifyBlocks                           | 13575           | 14588  | 14293  | 80372  | 301     |

### CODE REVIEW

**Review by davidtaikocha — APPROVED** [review]

File: packages/protocol/contracts/common/AddressResolver.sol [hunk 1]
```diff
@@ -72,11 +72,8 @@ abstract contract AddressResolver {
     function keyForName(
         uint256 chainId,
         string memory name
-    ) public pure returns (string memory key) {
-        key = string.concat(Strings.toString(chainId), ".", name);
-        // TODO: the next line is cheaper in gas but will break
-        //       many Hardhat tests.
-        // key = string(bytes.concat(bytes32(chainId), bytes(name)));
+    ) public pure virtual returns (string memory) {
+        return string(bytes.concat(bytes32(chainId), bytes(name)));
```
Comment (adaki2004) [inline packages/protocol/contracts/common/AddressResolver.sol:76]: Just a follow up on this mechanism:
https://github.com/taikoxyz/taiko-mono/pull/13285#discussion_r1136087299
We need to be sure that off-chain (tests or BE/FE client side typescript) we reconstruct it the same way.

**Review by adaki2004 — APPROVED** [review]
We need to be sure that off-chain (tests or BE/FE client side typescript) we reconstruct it the same way.

File: packages/protocol/contracts/common/AddressResolver.sol (hunk 1 above)
Comment (dantaik) [reply packages/protocol/contracts/common/AddressResolver.sol:76]: We do not use the key offchain in frontend or backend.

File: packages/protocol/test2/genesis/GenerateGenesis.t.sol [hunk 2]
```diff
@@ -174,12 +174,10 @@ contract TestGenerateGenesis is Test, AddressResolver {
     }
 
     function testERC20() public {
-        TestERC20 testErc20 = TestERC20(
-            getPredeployedContractAddress("TestERC20")
-        );
+        FooToken fooErc20 = FooToken(getPredeployedContractAddress("FooToken"));
```
Comment (dantaik) [inline packages/protocol/test2/genesis/GenerateGenesis.t.sol:177]: @davidtaikocha  use a new name for the regular erc20 token

**Pull request merged**